libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -97,321 +97,412 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
+    thrust_heading_rotation_angles(_attitude_target_quat, attitude_vehicle_quat, attitude_error_vector, _thrust_error_angle);
 
-    // Keep euler derivative updated
-    ang_vel_to_euler_rate(Vector3f(_ahrs.roll,_ahrs.pitch,_ahrs.yaw), _ang_vel_target_rads, _att_target_euler_rate_rads);
+    // Compute the angular velocity target from the attitude error
+    _rate_target_ang_vel = update_ang_vel_target_from_att_error(attitude_error_vector);
+
+    // Add feedforward term that attempts to ensure that roll and pitch errors rotate with the body frame rather than the reference frame.
+    _rate_target_ang_vel.x += attitude_error_vector.y * _ahrs.get_gyro().z;
+    _rate_target_ang_vel.y += -attitude_error_vector.x * _ahrs.get_gyro().z;
+
+    // Add the angular velocity feedforward, rotated into vehicle frame
+    Quaternion attitude_target_ang_vel_quat = Quaternion(0.0f, _attitude_target_ang_vel.x, _attitude_target_ang_vel.y, _attitude_target_ang_vel.z);
+    Quaternion attitude_error_quat = attitude_vehicle_quat.inverse() * _attitude_target_quat;
+    Quaternion target_ang_vel_quat = attitude_error_quat.inverse()*attitude_target_ang_vel_quat*attitude_error_quat;
+
+    // Correct the thrust vector and smoothly add feedforward and yaw input
+    if(_thrust_error_angle > AC_ATTITUDE_THRUST_ERROR_ANGLE*2){
+        _rate_target_ang_vel.z = _ahrs.get_gyro().z;
+    }else if(_thrust_error_angle > AC_ATTITUDE_THRUST_ERROR_ANGLE){
+        float flip_scalar = (1.0f - (_thrust_error_angle-AC_ATTITUDE_THRUST_ERROR_ANGLE)/AC_ATTITUDE_THRUST_ERROR_ANGLE);
+        _rate_target_ang_vel.x += target_ang_vel_quat.q2*flip_scalar;
+        _rate_target_ang_vel.y += target_ang_vel_quat.q3*flip_scalar;
+        _rate_target_ang_vel.z += target_ang_vel_quat.q4;
+        _rate_target_ang_vel.z = _ahrs.get_gyro().z*(1.0-flip_scalar) + _rate_target_ang_vel.z*flip_scalar;
+    } else {
+        _rate_target_ang_vel.x += target_ang_vel_quat.q2;
+        _rate_target_ang_vel.y += target_ang_vel_quat.q3;
+        _rate_target_ang_vel.z += target_ang_vel_quat.q4;
+    }
+
+    // rotate target and normalize
+    Quaternion attitude_target_update_quat;
+    attitude_target_update_quat.from_axis_angle(Vector3f(_attitude_target_ang_vel.x * _dt, _attitude_target_ang_vel.y * _dt, _attitude_target_ang_vel.z * _dt));
+    _attitude_target_quat = _attitude_target_quat * attitude_target_update_quat;
+    _attitude_target_quat.normalize();
 }
 
-void AC_AttitudeControl::input_att_quat_bf_ang_vel(const Quaternion& att_target_quat, const Vector3f& att_target_ang_vel_rads)
+// get_thrust_heading_rotation - calculates two ordered rotations to move the att_from_quat quaternion to the att_to_quat quaternion.
+// The first rotation corrects the thrust vector and the second rotation corrects the heading vector.
+void AC_AttitudeControl::thrust_heading_rotation_angles(Quaternion& att_to_quat, Quaternion att_from_quat, Vector3f& att_diff_angle, float& thrust_vec_dot)
```

Comment:
I am not sure how I should be doing this formally. I just figured if I didn't change it I shouldn't do it that way but maybe I should be making it a const or something.....
Guidance would be great here!!